File: sample/src/main/java/com/stepstone/stepper/sample/adapter/SampleStepAdapter.java
package com.stepstone.stepper.sample.adapter;

import android.content.Context;
import android.support.v4.view.ViewPager;
import android.util.SparseArray;
import android.view.View;
import android.view.ViewGroup;

import com.stepstone.stepper.Step;
import com.stepstone.stepper.adapter.AbstractStepAdapter;
import com.stepstone.stepper.sample.step.view.StepViewSample;

/**
 * A naive implementation of {@link AbstractStepAdapter}.
 * This does not keep the view data on rotation, etc.
 * It also keeps a reference to the created views.
 */
public class SampleStepAdapter extends AbstractStepAdapter {

    private final SparseArray<Step> pages = new SparseArray<>();

    public SampleStepAdapter(Context context) {
        super(context);
    }

    @Override
    public StepViewSample createStep(int position) {
        return new StepViewSample(context);
    }

    @Override
    public int getCount() {
        return 3;
    }

    @Override
    public Step findStep(ViewPager viewPager, int position) {
        return pages.size() > 0 ? pages.get(position) : null;
    }

    @Override
    public View instantiateItem(ViewGroup container, int position) {
        Step step = pages.get(position);
        if (step == null) {
            step = createStep(position);
            pages.put(position, step);
        }

        View stepView = (View) step;
        container.addView(stepView);

        return stepView;
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        container.removeView((View) object);
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return view == object;
    }
}

File: sample/src/main/java/com/stepstone/stepper/sample/adapter/SampleStepAdapter.kt
package com.stepstone.stepper.sample.adapter

import android.content.Context
import android.support.annotation.IntRange
import android.util.SparseArray
import android.view.View
import android.view.ViewGroup

import com.stepstone.stepper.Step
import com.stepstone.stepper.adapter.AbstractStepAdapter
import com.stepstone.stepper.sample.R
import com.stepstone.stepper.sample.step.view.StepViewSample
import com.stepstone.stepper.viewmodel.StepViewModel

/**
 * A naive implementation of [AbstractStepAdapter].
 * This does not keep the view data on rotation, etc.
 * It also keeps a reference to the created views.
 */
class SampleStepAdapter(context: Context) : AbstractStepAdapter(context) {

    private val pages = SparseArray<Step>()

    override fun createStep(position: Int): StepViewSample {
        return StepViewSample(context)
    }

    override fun getViewModel(@IntRange(from = 0) position: Int): StepViewModel {
        return StepViewModel.Builder(context)
                .setTitle(R.string.tab_title)
                .create()
    }

    override fun getCount(): Int {
        return 3
    }

    override fun findStep(position: Int): Step? {
        return if (pages.size() > 0) pages.get(position) else null
    }

    override fun instantiateItem(container: ViewGroup, position: Int): View {
        var step: Step? = pages.get(position)
        if (step == null) {
            step = createStep(position)
            pages.put(position, step)
        }

        val stepView = step as View
        container.addView(stepView)

        return stepView
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        container.removeView(`object` as View)
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === `object`
    }
}
